src/main/java/se/liu/ida/hefquin/engine/query/impl/BGPImpl.java

```diff
@@ -30,7 +32,19 @@ public boolean equals( final Object o ) {
 			otps = ((BGPImpl) o).tps;
 		else
 			otps = ((BGP) o).getTriplePatterns();
-		if ( tps == otps )
+		
+		final Set<Triple> tpsTriple = new HashSet<>();
+		final Set<Triple> otpsTriple = new HashSet<>();
+		
+		for(TriplePattern i : tps) {
+			tpsTriple.add(i.asJenaTriple());
+		}
+		
+		for(TriplePattern j : otps) {
+			otpsTriple.add(j.asJenaTriple());
+		}
+		
```

Comment:
Why not having sets of `TriplePattern`??

Reply:
I realised that when I compared sets the equals function of the objects in the sets where never called. So The TriplePatterns are not considered equal in the sets, but the jena triples are. I don't know if there's any way around this? Lists disn't work because of the order.

Reply:
> The TriplePatterns are not considered equal in the sets, but the jena triples are.

That's strange! The `equals` method in `TriplePatternImpl` is implemented correctly (by relying on `equals` of the underlying Jena `Triple` objects). Perhaps our assumptions about `equals` of Java collection types (such as `Set`) are too strong. Maybe they do not check containment based on the `equals` function of the contained objects but just based on the `==` operator?